File: aula09-18/UTFQuestPolimorfismo.java
/*
* UTF Quest: um RPG em modo texto
*
* Material complementar da aula "Polimorfismo"
* Disciplina: Programação Orientada a Objetos (UTFPR, Campus Apucarana)
* Professor: Luiz Rodrigues (luizrodrigues@utfpr.edu.br)
*
* Observações:
* - Este arquivo reúne as classes para facilitar o compartilhamento.
*   Em projetos maiores, cada classe pública vive em seu próprio arquivo.
* - Evolução em relação à aula passada: o UTF Quest ganhou polimorfismo.
*   Um objeto da subclasse agora é tratado pelo tipo da superclasse
*   (upcasting), e a ligação dinâmica garante que a versão executada de
*   um método sobrescrito é sempre a do objeto real. A lista de
*   Personagem reúne toda a família, e o laço da batalha fala um só
*   comando: cada herói responde do seu jeito.
* - Personagem virou uma classe abstrata: no jogo, todo herói pertence
*   a uma classe concreta, e new Personagem(...) deixou de fazer
*   sentido. O método abstrato habilidade() obriga cada subclasse a
*   ter a sua.
* - O Chefe é a prova do código aberto para crescer: uma subclasse
*   nova, criada com o construtor de três argumentos da aula de
*   sobrecarga, e o laço da batalha não mudou uma linha.
* - instanceof e o cast (downcasting) aparecem uma única vez, para o
*   que é só do Mago: no dia a dia, prefira sobrescrever o método.
*
* Como executar (JDK 11 ou superior):
*   java UTFQuestPolimorfismo.java
*/

import java.util.ArrayList;

public class UTFQuestPolimorfismo {

  public static void main(String[] args) {

    // Upcasting: um Mago cabe numa variável Personagem.
    // Ligação dinâmica: o ficha() executado é o do Mago.
    Personagem p = new Mago("Elara");
    System.out.println(p.ficha());
    System.out.println();

    // Uma lista para todos os heróis: coleção polimórfica
    ArrayList<Personagem> herois = new ArrayList<>();
    herois.add(new Mago("Elara"));
    herois.add(new Guerreiro("Bran"));

    // Um comando, vários comportamentos
    for (Personagem h : herois) {
      System.out.println(h.ficha());
    }
    System.out.println();

    // instanceof descobre o tipo real...
    for (Personagem h : herois) {
      if (h instanceof Mago) {
        System.out.println(h.getNome() + " é mago!");
      }
    }
    // ...e o cast (downcasting) libera o que é só do Mago
    for (Personagem h : herois) {
      if (h instanceof Mago) {
        Mago m = (Mago) h;
        System.out.println(m.getMana());
      }
    }
    System.out.println();

    // A batalha final: o laço não sabe, nem precisa saber,
    // a classe de cada herói
    Chefe chefe = new Chefe("Rei Lich");

    for (Personagem h : herois) {
      System.out.println(h.getNome() + " usa " + h.habilidade());
      h.atacar(chefe);
    }
    System.out.println();

    // O chefe revida: o receberDano() do Guerreiro desconta a
    // defesa, mesmo com Bran visto como Personagem
    for (Personagem h : herois) {
      chefe.atacar(h);
    }
    System.out.println();

    for (Personagem h : herois) {
      System.out.println(h.ficha());
    }
    System.out.println(chefe.ficha());
  }
}

// Classe abstrata: não gera objetos; existe para as filhas herdarem
abstract class Personagem {

  // Atributos private; valores fixos inicializados na declaração
  private String nome;
  private int vida = 100;
  private int forca = 10;
  private ArrayList<Item> inventario = new ArrayList<>();

  // Construtores sobrecarregados: duas formas de nascer

  public Personagem(String nome) {
    setNome(nome);
  }

  public Personagem(String nome, int vida, int forca) {
    this(nome);   // encadeia: chama o construtor acima
    setVida(vida);
    setForca(forca);
  }

  // Método abstrato: cada subclasse escreve o seu
  abstract String habilidade();

  // Getters: leitura controlada

  public String getNome() {
    return nome;
  }

  public int getVida() {
    return vida;
  }

  public int getForca() {
    return forca;
  }

  // Setters: escrita validada; o objeto protege o próprio estado

  public void setNome(String n) {
    if (n != null && !n.isEmpty()) {
      nome = n;
    } else {
      System.out.println("nome inválido");
    }
  }

  public void setVida(int v) {
    if (v >= 0) {
      vida = v;
    } else {
      System.out.println("vida inválida: " + v);
    }
  }

  public void setForca(int f) {
    if (f >= 0 && f <= 100) {
      forca = f;
    } else {
      System.out.println("força inválida: " + f);
    }
  }

  // Interface pública: as ações que o jogo pode usar

  void receberDano(int dano) {
    setVida(Math.max(0, vida - dano));
    System.out.println(nome + " sofreu " + dano + " de dano");
  }

  // Sobrecarga de métodos: mesmo nome, assinaturas diferentes

  void curar() {
    curar(10);
  }

  void curar(int qtd) {
    setVida(vida + qtd);
  }

  boolean estaVivo() {
    return vida > 0;
  }

  String ficha() {
    return nome + " (vida: " + vida + ", forca: " + forca + ")";
  }

  // Parâmetro polimórfico: qualquer personagem serve como alvo
  void atacar(Personagem alvo) {
    System.out.println(nome + " ataca " + alvo.getNome());
    alvo.receberDano(forca);
  }

  void pegar(Item i) {
    inventario.add(i);
    setForca(forca + i.getBonusForca());
    System.out.println(nome + " pegou " + i.getNome() + " (forca agora: " + forca + ")");
  }
}

class Item {

  private String nome;
  private int bonusForca;

  // O Item também exige os dados essenciais para nascer
  Item(String nome, int bonus) {
    setNome(nome);
    setBonusForca(bonus);
  }

  public String getNome() {
    return nome;
  }

  public int getBonusForca() {
    return bonusForca;
  }

  public void setNome(String n) {
    nome = n;
  }

  public void setBonusForca(int b) {
    if (b >= 0) {
      bonusForca = b;
    } else {
      System.out.println("bônus inválido: " + b);
    }
  }
}

class Mago extends Personagem {

  // Só o que é próprio do Mago; o resto vem por herança
  private int mana;

  public Mago(String nome) {
    super(nome);      // Personagem inicializa nome, vida e forca
    this.mana = 50;   // depois, a parte que é só do Mago
  }

  public int getMana() {
    return mana;
  }

  // Obrigação cumprida: a habilidade do Mago
  @Override
  String habilidade() {
    return "rajada arcana";
  }

  // Sobrescrita: mesmo nome, mesma assinatura, corpo novo.
  // super.ficha() reaproveita a versão herdada e apenas completa.
  @Override
  String ficha() {
    return super.ficha() + " [mana: " + mana + "]";
  }
}

class Guerreiro extends Personagem {

  private int defesa;

  public Guerreiro(String nome) {
    super(nome);
    this.defesa = 5;
  }

  public int getDefesa() {
    return defesa;
  }

  @Override
  String habilidade() {
    return "escudo de aço";
  }

  // Sobrescrita: a defesa absorve parte do dano antes de
  // delegar ao comportamento herdado do Personagem
  @Override
  void receberDano(int dano) {
    super.receberDano(Math.max(0, dano - defesa));
  }
}

// A subclasse nova entra na família sem mudar o laço da batalha
class Chefe extends Personagem {

  public Chefe(String nome) {
    super(nome, 300, 20);   // o construtor de três argumentos
  }

  @Override
  String habilidade() {
    return "golpe devastador";
  }
}

File: aula09-18/utf_quest_polimorfismo.py
# UTF Quest: um RPG em modo texto
#
# Material complementar da aula "Polimorfismo"
# Disciplina: Programação Orientada a Objetos (UTFPR, Campus Apucarana)
# Professor: Luiz Rodrigues (luizrodrigues@utfpr.edu.br)
#
# Observações:
# - Evolução em relação à aula passada: o UTF Quest ganhou polimorfismo.
#   Em Python as variáveis não declaram tipo, então a lista mista sempre
#   funcionou; a herança é o que garante que todos na lista respondem
#   aos mesmos comandos, cada um com a sua versão (ligação dinâmica).
# - Personagem virou uma classe abstrata (ABC): no jogo, todo herói
#   pertence a uma classe concreta, e Personagem(...) agora levanta
#   TypeError. O método abstrato habilidade() obriga cada subclasse a
#   ter a sua.
# - O Chefe é a prova do código aberto para crescer: uma subclasse
#   nova, criada com os parâmetros extras do __init__ (a "sobrecarga"
#   à moda Python), e o laço da batalha não mudou uma linha.
# - isinstance aparece uma única vez, para o que é só do Mago: no dia
#   a dia, prefira sobrescrever o método.
#
# Como executar (Python 3):
#   python utf_quest_polimorfismo.py

from abc import ABC, abstractmethod


class Personagem(ABC):
  # Classe abstrata: não gera objetos; existe para as filhas herdarem

  # Parâmetros padrão fazem o papel da sobrecarga de construtores
  def __init__(self, nome, vida=100, forca=10):
    self._nome = ""
    self._vida = 100
    self._forca = 10
    self._inventario = []
    self.nome = nome      # passa pelas properties: nasce validado
    self.vida = vida
    self.forca = forca

  # Método abstrato: cada subclasse escreve o seu
  @abstractmethod
  def habilidade(self):
    ...

  # Properties: leitura e escrita controladas com sintaxe natural

  @property
  def nome(self):
    return self._nome

  @nome.setter
  def nome(self, n):
    if n:
      self._nome = n
    else:
      print("nome inválido")

  @property
  def vida(self):
    return self._vida

  @vida.setter
  def vida(self, v):
    if v >= 0:
      self._vida = v
    else:
      print("vida inválida:", v)

  @property
  def forca(self):
    return self._forca

  @forca.setter
  def forca(self, f):
    if 0 <= f <= 100:
      self._forca = f
    else:
      print("força inválida:", f)

  # Interface pública: as ações que o jogo pode usar

  def receber_dano(self, dano):
    self.vida = max(0, self.vida - dano)
    print(self.nome, "sofreu", dano, "de dano")

  # Parâmetro padrão faz o papel da sobrecarga de métodos
  def curar(self, qtd=10):
    self.vida = self.vida + qtd

  def esta_vivo(self):
    return self.vida > 0

  def ficha(self):
    return (self.nome + " (vida: " + str(self.vida) + ", forca: " + str(self.forca) + ")")

  # Parâmetro polimórfico: qualquer personagem serve como alvo
  def atacar(self, alvo):
    print(self.nome, "ataca", alvo.nome)
    alvo.receber_dano(self.forca)

  def pegar(self, i):
    self._inventario.append(i)
    self.forca = self.forca + i.bonus_forca
    print(self.nome, "pegou", i.nome, "(forca agora: " + str(self.forca) + ")")


class Item:

  # O Item também exige os dados essenciais para nascer
  def __init__(self, nome, bonus):
    self._nome = ""
    self._bonus_forca = 0
    self.nome = nome
    self.bonus_forca = bonus

  @property
  def nome(self):
    return self._nome

  @nome.setter
  def nome(self, n):
    self._nome = n

  @property
  def bonus_forca(self):
    return self._bonus_forca

  @bonus_forca.setter
  def bonus_forca(self, b):
    if b >= 0:
      self._bonus_forca = b
    else:
      print("bônus inválido:", b)


class Mago(Personagem):

  # Só o que é próprio do Mago; o resto vem por herança
  def __init__(self, nome):
    super().__init__(nome)   # Personagem inicializa nome, vida e forca
    self._mana = 50          # depois, a parte que é só do Mago

  @property
  def mana(self):
    return self._mana

  # Obrigação cumprida: a habilidade do Mago
  def habilidade(self):
    return "rajada arcana"

  # Sobrescrita: mesmo nome, mesma assinatura, corpo novo.
  # super().ficha() reaproveita a versão herdada e apenas completa.
  def ficha(self):
    return super().ficha() + " [mana: " + str(self._mana) + "]"


class Guerreiro(Personagem):

  def __init__(self, nome):
    super().__init__(nome)
    self._defesa = 5

  @property
  def defesa(self):
    return self._defesa

  def habilidade(self):
    return "escudo de aço"

  # Sobrescrita: a defesa absorve parte do dano antes de
  # delegar ao comportamento herdado do Personagem
  def receber_dano(self, dano):
    super().receber_dano(max(0, dano - self._defesa))


class Chefe(Personagem):
  # A subclasse nova entra na família sem mudar o laço da batalha

  def __init__(self, nome):
    super().__init__(nome, 300, 20)   # os parâmetros extras

  def habilidade(self):
    return "golpe devastador"


if __name__ == "__main__":

  # Em Python p aceita qualquer objeto; a ligação dinâmica garante
  # que o ficha() executado é o do Mago
  p = Mago("Elara")
  print(p.ficha())
  print()

  # Uma lista para todos os heróis: coleção polimórfica
  herois = [Mago("Elara"), Guerreiro("Bran")]

  # Um comando, vários comportamentos
  for h in herois:
    print(h.ficha())
  print()

  # isinstance descobre o tipo real...
  for h in herois:
    if isinstance(h, Mago):
      print(h.nome, "é mago!")
  # ...e aqui nenhuma conversão é necessária: Python olha o objeto
  for h in herois:
    if isinstance(h, Mago):
      print(h.mana)
  print()

  # A batalha final: o laço não sabe, nem precisa saber,
  # a classe de cada herói
  chefe = Chefe("Rei Lich")

  for h in herois:
    print(h.nome, "usa", h.habilidade())
    h.atacar(chefe)
  print()

  # O chefe revida: o receber_dano() do Guerreiro desconta a
  # defesa, mesmo com Bran tratado como um personagem qualquer
  for h in herois:
    chefe.atacar(h)
  print()

  for h in herois:
    print(h.ficha())
  print(chefe.ficha())
